### PR TITLE
Flexbe hfsmbth v2

### PR DESCRIPTION
**Pull request by dcconner**
- update license
- add demo of the "mythical HFSMBTH" using BT.cpp
  - Required change to `pyrobosim_btcpp` to export headers

### CODE REVIEW

**Comment by dcconner** [comment]
@facontidavide I asked for your review mainl because this requires a change to your CMakeLists.txt (included)

You may have input on how I implemented the custom BT executor.
(It does include some unnecessarily verbose status message for demo.)

Additionally, when I was trying to run with Groot2 in the Docker container, the default port is 5555 on startup, where the default port in code was 1667 (which I think is your standard).  

What's up with that?
I think for shop we should match default.

Also, I'm getting 

```
"Reply timeout"
socket disconnected
Warning: Tue Oct 8 02:35:03 2024 - Disconnected with error:  "Reply timeout" ,
 Reconnecting...
subscribe finished
Info: Tue Oct 8 02:35:03 2024 - "Error loading  model" "The model 'LoopDouble' is builtin model and is being overriden...."
Info: Tue Oct 8 02:35:03 2024 - "Error loading  model" "The model 'LoopString' is builtin model and is being overriden...."
Info: Tue Oct 8 02:35:03 2024 - "Error loading  model" "The model 'Script' is builtin model and is being overriden...."
Info: Tue Oct 8 02:35:03 2024 - "Error loading  model" "The model 'ScriptCondition' is builtin model and is being overriden...."
Info: Tue Oct 8 02:35:03 2024 - "Error loading  model" "The model 'SetBlackboard' is builtin model and is being overriden...."
Info: Tue Oct 8 02:35:03 2024 - "Error loading  model" "The model 'Switch2' is builtin model and is being overriden...."
Info: Tue Oct 8 02:35:03 2024 - "Error loading  model" "The model 'Switch3' is builtin model and is being overriden...."
Info: Tue Oct 8 02:35:03 2024 - "Error loading  model" "The model 'Switch4' is builtin model and is being overriden...."
Info: Tue Oct 8 02:35:03 2024 - "Error loading  model" "The model 'Switch5' is builtin model and is being overriden...."
Info: Tue Oct 8 02:35:03 2024 - "Error loading  model" "The model 'Switch6' is builtin model and is being overriden...."
```

which I don't recall seeing in earlier testing.

File: technologies/FlexBE/README.md [hunk 1]
```diff
@@ -89,6 +90,7 @@ The `pyrobosim_flexbe_states` package under `technologies/FlexBE` includes:
 * `pick_action_state.py` - Does pick action at current location
 * `place_action_state.py` - Does place action at current location
 * `plan_path_state.py` - Request plan from current location to target location
+* `run_btcpp_tree_state.py` - Run a behavior tree using custom BehaviorTree.CPP v4 executor
```
Comment (sea-bass) [inline technologies/FlexBE/README.md:93]: IMO not needed to specify `v4`, as the convention is that is the latest, and if you're on the older versions you specify it.

File: technologies/FlexBE/docs/hfsmbth.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
+# The Mythical HFSMBTH
```
Comment (sea-bass) [inline technologies/FlexBE/docs/hfsmbth.md:1]: Might want to fully define this once as Hierarchical Finite-State Machine / Behavior Tree Hybrid.

File: technologies/FlexBE/docs/hfsmbth.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+Inspired by the 2019 Game Developers Conference [talk](https://www.youtube.com/watch?v=Qq_xX1JCreI&t=1159s) by Bobby Anguelov, we present the "Mythical HFSMBT Hybrid".
+
+Please use the following publication for reference:
+
+* Joshua M. Zutell, David C. Conner, and Philipp Schillinger, "Flexible Behavior Trees: In search of the mythical HFSMBTH for Collaborative Autonomy in Robotics", arXiv 2022, <https://doi.org/10.48550/arXiv.2203.05389>.
+
+## BehaviorTree.cpp Hybrid Demo
```
Comment (sea-bass) [inline technologies/FlexBE/docs/hfsmbth.md:9]: following official tool name styling

```suggestion
## BehaviorTree.CPP Hybrid Demo

File: technologies/FlexBE/docs/hfsmbth.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+* `clear; ros2 run delib_ws_worlds run --ros-args -p problem_number:=2`
+* `clear; ros2 launch flexbe_onboard behavior_onboard.launch.py`
+* `clear; ros2 launch flexbe_webui flexbe_ocs.launch.py headless:=true`
+* `clear; ros2 run flexbe_webui webui_client`
+
+Additionally we need to launch the custom `flexbe_btcpp_executor` node which executes the
+BT.cpp v4-based behavior trees using an `ExecuteTree` action server interface.
```
Comment (sea-bass) [inline technologies/FlexBE/docs/hfsmbth.md:17]: same comments as before, you can say BT.CPP (all caps) and omit the v4

File: technologies/FlexBE/pyrobosim_flexbe_btcpp/config/pyrobosim_flexbe_btcpp.yaml [hunk 5]
```diff
@@ -0,0 +1,13 @@
+bt_action_server:
+  ros__parameters:
+    action_name: "flexbe_bt_server" # Optional (defaults to `bt_action_server`)
+    tick_frequency: 100 # Optional (defaults to 100 Hz)
+    groot2_port: 1667 # Optional (defaults to 1667)
+    ros_plugins_timeout: 1000  # Optional (defaults 1000 ms)
+
+#    plugins:
+#      - behaviortree_cpp/bt_plugins
+#      - btcpp_ros2_samples/bt_plugins
```
Comment (sea-bass) [inline technologies/FlexBE/pyrobosim_flexbe_btcpp/config/pyrobosim_flexbe_btcpp.yaml:10]: can this be removed if not used?

File: technologies/FlexBE/pyrobosim_flexbe_btcpp/src/flexbe_btcpp_executor.cpp [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+  */
+  std::optional<std::string> onLoopFeedback()
+  {
+    return std::nullopt;
+  }
+
+private:
+  std::shared_ptr<BT::StdCoutLogger> logger_cout_;
+  rclcpp::Subscription<pyrobosim_msgs::msg::RobotState>::SharedPtr sub_;
+  uint32_t tic_count_;
```
Comment (sea-bass) [inline technologies/FlexBE/pyrobosim_flexbe_btcpp/src/flexbe_btcpp_executor.cpp:197]: nit: the term used is "tick", so would prefer `tick_count_`.

File: technologies/FlexBE/pyrobosim_flexbe_btcpp/src/flexbe_btcpp_executor.cpp [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+   * If you don't want to return any feedback, return std::nullopt.
+  */
+  std::optional<std::string> onLoopFeedback()
+  {
+    return std::nullopt;
+  }
+
+private:
+  std::shared_ptr<BT::StdCoutLogger> logger_cout_;
+  rclcpp::Subscription<pyrobosim_msgs::msg::RobotState>::SharedPtr sub_;
```
Comment (sea-bass) [inline technologies/FlexBE/pyrobosim_flexbe_btcpp/src/flexbe_btcpp_executor.cpp:196]: nit: personal preference, but I like to call these variables by what they subscribe to, e.g., `state_sub_` or `robot_state_sub_`.

File: technologies/FlexBE/pyrobosim_flexbe_btcpp/trees/problem1.xml [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+        <IfThenElse>
+            <DetectObject name="DetectSnack" object="snacks0"/>
+            <Sequence name="DetectedTrue">
+                <PickObject name="PickSnack" object="snacks0"/>
+                <Navigate name="ToTableWithSnack" target="table"/>
+                <PlaceObject name="PlaceSnack" object="snacks0"/>
+            </Sequence>
+            <Navigate name="ToTableWithoutSnack" target="table"/>
+        </IfThenElse>
+        <Sequence name="DetecSnackBackToPantry">
```
Comment (sea-bass) [inline technologies/FlexBE/pyrobosim_flexbe_btcpp/trees/problem1.xml:15]: I think I had this comment in the original BT.CPP PR, but

```suggestion
        <Sequence name="DetectSnackBackToPantry">

File: technologies/FlexBE/docs/hfsmbth.md [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+* `clear; ros2 launch pyrobosim_flexbe_btcpp pyrobosim_flexbe_btcpp.launch.xml`
+
+This custom executor node loads the node definitions used by this workshop demonstration.
+
+Via the FlexBE webui load the `BtCppHFSMBTH` behavior and execute.
+    This uses a `RunBtCppTreeState` to send `ExecuteTree` action goals to the custom `flexbe_btcpp_executor` node.
```
Comment (sea-bass) [inline technologies/FlexBE/docs/hfsmbth.md:24]: Is this extra spacing at the start of the line intentional?

File: technologies/FlexBE/docs/hfsmbth.md [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+```
+
+which is required if they are launched by the `ExecuteTree.action`.
+
+Additional behavior tree packages may be listed in the `pyrobosim_flexbe_btcpp/config/pyrobosim_flexbe_btcpp.yaml` configuration file, but the required node definitions must be
+registered with the `flexbe_btcpp_executor` code.
+
+The FlexBE operator may preempt the `RunBtCppTreeState` which sends a message to cancel the `ExecuteTree` action to the BT action server, which halts the behavior tree.
+
+You may optionally start `groot2` to visualize the behavior tree as it executes.
```
Comment (sea-bass) [inline technologies/FlexBE/docs/hfsmbth.md:39]: suggest showing the command

```bash
./Groot2.AppImage
```

**Review by sea-bass — COMMENTED** [review]
This is extremely cool!

Left some small comments, but will let @facontidavide chime in with the modifications and give the final approve.

image

**Comment by dcconner** [comment]
thanks for comments; cleaning them up was more fun than grading my AI midterm :-)

File: technologies/FlexBE/pyrobosim_flexbe_btcpp/src/flexbe_btcpp_executor.cpp [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+  */
+  void registerNodesIntoFactory(BT::BehaviorTreeFactory& factory)
+  {
+    RCLCPP_INFO(kLogger, "registerNodesIntoFactory - load custom nodes ...");
+    // all the actions are done using the same Action Server.
+    // Therefore  single RosNodeParams will do.
+    BT::RosNodeParams params;
+    params.nh = this->node();
+    params.default_port_value = "execute_action";
+
```
Comment (facontidavide) [inline technologies/FlexBE/pyrobosim_flexbe_btcpp/src/flexbe_btcpp_executor.cpp:159]: you might wany to add  `IsBatteryLow` and `IsBatteryFull`, too

**Comment by facontidavide** [comment]
very nice! just a minor comment!

File: technologies/FlexBE/pyrobosim_flexbe_btcpp/config/pyrobosim_flexbe_btcpp.yaml [hunk 12]
```diff
@@ -0,0 +1,9 @@
+bt_action_server:
+  ros__parameters:
+    action_name: "flexbe_bt_server" # Optional (defaults to `bt_action_server`)
+    tick_frequency: 100 # Optional (defaults to 100 Hz)
+    groot2_port: 1667 # Optional (defaults to 1667)
```
Comment (facontidavide) [inline technologies/FlexBE/pyrobosim_flexbe_btcpp/config/pyrobosim_flexbe_btcpp.yaml:5]:
```suggestion
    groot2_port: 5555
```

Change to use the same default as Groot2

**Review by facontidavide — CHANGES_REQUESTED** [review]

**Comment by facontidavide** [comment]
image

**Comment by dcconner** [comment]
addressed @facontidavide comments and had to move BehaviorTree.CPP to latest humble branch that includes fix to issue that is required for this demo.

Apparently a merge undid that prior fix.

**Review by sea-bass — APPROVED** [review]
The comments all seem to be addressed ✅

**Pull request merged**